### ISSUE TITLE
Location Accuracy stuck at City, can't get neighborhood.

### ISSUE THREAD

**Issue by DancinDirk** [intuit/LocationManager#87]
Recently I began working on an app for a client that was using INTULocationManager 4.1.1. The first few updates went well and I didn't have to touch the Location code at all but recently we've been having an issue. We're asking for .Neighborhood level accuracy and after doing some debugging we're finding that location is always showing accuracy is at .City after timing out. Tried setting the timeout obscenely high and walking around with the app but it never get's better than .City. If I drop the accuracy requested to .City the rest of the app functions normally although less accurately. I really need better accuracy than .City so this is only a temporary workaround. Might be worth noting that the issue started around the iOS 10 release.

Any and all help is much appreciated!

**Comment by Bwaldrop3282** [comment]
I am having the same issue, I am also seeing it work fine on some devices and not on others. For instance, on a 6S, no issues. on a 6, doesn't ever get a GPS lock for .Neighborhood level accuracy. Works on 4S with iOS 9.x, no issues. 

I agree that iOS 10 is when the issue started.

**Comment by blueraz0r** [comment]
Could be an Apple bug with iOS 10. What is the currentLocation you get in the block when the location request times out?

**Comment by DancinDirk** [comment]
Current possible is the lowest level, city. Never gets better. An odd fact,
if you open a different location based app first, like a weather app, you
get a much better location. This is probably because the more accurate it
gets the shorter time and opening one immediately before stores a location
within that timeout. Not sure why location manager is getting a different
location level than other apps.

Thoughts?

**Comment by blueraz0r** [comment]
What does the `po currentLocation` in lldb debugger print? If you request a more accurate location does that help?

**Comment by DancinDirk** [comment]
Unless I'm misunderstanding the question, I was seeing Location City when printing it out. I was asking for Neighborhood at the time I noticed that it was only returning City. Asking for more detailed location did not help.

**Comment by blueraz0r** [comment]
What is the description of the CLLocation that you get? If you print the `currentLocation` in the debugger it will show when it was received and the accuracy in meters and more. Sharing that might help.

**Comment by valabal** [comment]
yeah me to.. I'm using ios 10.1 right now (my device is ipad air 2) cant get location when accuracy less then City/Neighboor.. (even I use the demo files in your github files its still not getting any result if I set accuracy under than city) 
my currentLocation is : 
<-6.90338116,+107.59653779> +/- 165.00m (speed -1.00 mps / course -1.00) @ 12/14/16, 10:57:31 PM Western Indonesia Time
(Its show that the accuracy is +/- 165.00m .. But when i set my accuracy to Block its always failed (either request time out (set it to 30 sec but its still time out) or error

**Comment by blueraz0r** [comment]
@valabal if the accuracy is +/- 165.00m that means it would be neighborhood. If it's less than 100m it would be block. https://github.com/intuit/LocationManager/blob/master/LocationManager/INTULocationManager/INTULocationRequestDefines.h#L56

**Comment by DancinDirk** [comment]
Here's what we get. Shows City level, accuracy around +/-1400m

Ran the app, hit the break point, got the error dialog from that case, pulled to refresh to attempt location a second time, same result. Tried this several times and never got a location. I also tried it with a 90 second timeout (not pictured) and that too comes up with the same result.

Thoughts?

![screenshot 2016-12-14 15 31 06](https://cloud.githubusercontent.com/assets/414755/21199669/77f141de-c212-11e6-93d5-2c9d55699f00.png)

**Comment by valabal** [comment]
@blueraz0r yeah I know, what I mean is everytime I use block accuracy its always prompt an error result message ( using the example project, everytime I choose to use Block/House/Room accuracy its always prompt @"An unknown error occurred.\n(Are you using iOS Simulator with location set to 'None'?)" 
I can't get success status when I use these accuracy no matter. (sometime when I choose Neighborhood its prompt an error result too)..
Maybe its bug in ios 10? (because when I open apple map application, the accuracy its not good either)...

**Comment by DancinDirk** [comment]
@blueraz0r Any further thoughts on this?
@lwdupont I see that you assigned the issue to yourself a couple weeks ago, do you have any insight or could I provide you with anything else that might help diagnose this issue?

Thanks in advance!

**Comment by DancinDirk** [comment]
Any movement on this issue at all? Should I start looking at other location Manager projects to replace this one? Anyone have experience with other projects similar to this?

**Comment by lwdupont** [comment]
Sorry, I've been very lax on taking care of this. Trying to recruit a few more folks to help as well. I will look at this today/this weekend and see what's up.

**Comment by DancinDirk** [comment]
Awesome, looking forward to what you find, thanks!

**Comment by lwdupont** [comment]
I suppose the good news is I can reproduce it with the test app that comes with the library. Still looking as to what's happening.

**Comment by lwdupont** [comment]
So, did some more investigation and testing with my iPhone 7Plus:

- inside my building, I can only ever get down to 65m accuracy, so asking for neighborhood works for me. 
- if I go to house (15m) or room (5)m, my request always times out. 
- if I walk outside as I'm doing the request for either, the phone immediately gets down to a 5m accuracy and returns a result. 

I haven't yet noticed any issues with the code in the framework. 

@DancinDirk in your screenshot above, what is the error that the framework returns to you? 

It's sounds like you were testing outside, as well, right?

I'm running on 10.2.1, the latest beta FWIW.

**Comment by DancinDirk** [comment]
@lwdupont We were getting timeout errors and we were testing in a mix of outside and in both with hit or miss results on a mixture of devices. 

For the most part it would seem to be using the last location used and not getting any better as indicated by the "opening weather app first" note in my second comment. It was quite repeatable to reboot the device (clearing any previous location data) and then open the app and have it time out. Then open a weather app and then switch back to our app and have a much better location. The question there is, how is a weather app able to get a better location fix while in the exact same location than we can while using our app with this library? The opening another app first is very puzzling.

**Comment by blueraz0r** [comment]
This might be an Apple bug. Or maybe due to changing the CLLocationManager desiredAccuracy here? https://github.com/intuit/LocationManager/blob/master/LocationManager/INTULocationManager/INTULocationManager.m#L480

**Comment by DancinDirk** [comment]
@lwdupont Have you had any additional luck with this?

**Comment by lwdupont** [comment]
Sorry @DancinDirk nothing yet, actually missed that you replied. I'll check on that code @blueraz0r . I suspect it has something to do with the code changing the accuracy returned based upon the last time it received it from the Apple api. 

@DancinDirk the timeout errors - did you set a timeout, or just ask for a location with no timeout?

**Comment by lwdupont** [comment]
I'm also enlisted my coworker @eoghain to help.

**Comment by DancinDirk** [comment]
@lwdupont Sorry I missed this question over the weekend.

I experimented with different timeout values, the standard was 10 sec, it was when we increased it to 90 that we noticed that we never got a more accurate location, I even tried 900 to see if it ever got better with no luck.

**Comment by DancinDirk** [comment]
Any updates @lwdupont ?

**Comment by lwdupont** [comment]
Actually @DancinDirk going to look at again it this afternoon/evening. I haven't found any quick fix yet though.

**Comment by DancinDirk** [comment]
Appreciate it @lwdupont, any initial thoughts on what could be the cause?

**Comment by DancinDirk** [comment]
Checking in again since it's been a couple weeks, any progress @lwdupont ?

**Comment by lwdupont** [comment]
No luck yet @DancinDirk. I've been slowly updating the code and test app to make testing easier.

**Comment by DancinDirk** [comment]
any updates @lwdupont ?

**Comment by DancinDirk** [comment]
Any updates on this issue? @lwdupont

**Comment by lwdupont** [comment]
Still poking at it. Stay tuned. (Busy tax season, but that's slowing down now)

**Comment by buithuyen** [comment]
I have the same one. Any update for that?

**Comment by lwdupont** [comment]
Sorry for the lack of communication here. I think I have a decent idea of the issue, but I could really use some info from anyone who has this issue:

- the exact code you are calling on this framework, how you are setting it up, and what you are expecting result wise. 

Feel free to email it to lucien_dupont at intuit dot com. I'm not providing email support, but before I go and rewrite some stuff I really want to have something reproducible in an app outside of the example provided in here. 

Thanks. cc @DancinDirk @buithuyen

**Comment by lwdupont** [comment]
Spent a few hours tonight writing a new app in Swift with Xcode 8 using the currently released INTULocationManager pod. I copied @DancinDirk 's code from his screenshot basically, added a small UI to it. Basically to try and prove what I thought was the issue (there's some code when requesting a one time location that expires the current location if it hasn't been updated within a certain number of seconds, less seconds for the more accurate location request). 

When I set my request to house and 30 seconds, inside my house I timed out. When I ran it and walked outside, it within a second or two found an accurate location. 

Here's the source: https://github.com/lwdupont/TestIntuitLocation

Basically what I was expecting. 

So I decided to run the example app I've updated on the Code_cleanup branch here that I've been working on. Similar issue when sitting inside my house. However, when I walked outside with a request going, it never got a lock and timed out. Multiple times. 

I thought it might be the fact that example app's deployment target was 9.0 (and my example program was 10.3), but I changed mine to 9.0, and got the same results. 

Makes me think that the issue is with the example program. Although it doesn't discount @DancinDirk issue. 

So, I'm going to double check the example program more, but regarding others issues, I haven't been able to prove to myself that there is an issue in the library, and hence can't start to look for the problem if I can't reproduce it. 

I'll close this in a few days if I don't hear from anyone else on the topic. 

PS: Will try to release a new version of this pod by next weekend.
